### PR TITLE
Add support for ZoneInfo

### LINKED ISSUE #4710: Add timezone to tzinfo parsing to pydantic 2.0
### Discussed in https://github.com/pydantic/pydantic/discussions/4708

<div type='discussions-op-text'>

<sup>Originally posted by **07pepa** November  2, 2022</sup>
Pydantic 1.x is missing native suport for timezone parsing  (and since this will probably will not be added natively). I am asking to add it to pydantic 2.x (does not have to be in rust) 

hopefully there is something like pytz (but there is issue how to keep database of all timezone up to date without need to recompile/ update rust part) 

</div>

### PR DESCRIPTION
**Pull request by Youssefares**
## Related issue number
fix #4710 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @sydney-runkle

### CODE REVIEW

**Comment by yezz123** [comment]
hey @Youssefares correct me if i'm wrong but i guess its similar to this one we have in Pydantic-extra https://github.com/pydantic/pydantic-extra-types/pull/193

**Comment by Youssefares** [comment]
> hey @Youssefares correct me if i'm wrong but i guess its similar to this one we have in Pydantic-extra [pydantic/pydantic-extra-types#193](https://github.com/pydantic/pydantic-extra-types/pull/193)

This does seem very similar, zoneinfo.ZoneInfo is part of the python std library though. I maybe wrong but my understanding of pydantic-extra-types is that it defines additional convenience types, while pydantic itself is where we offer support for standard types. I am not 100% sure about the need for an extra-type in this scenario given zoneinfo now exists since 3.9.

I'll leave to maintainers to decide whether the extra type using pytz is still required, but I believe supporting zoneinfo is very much worthwhile eitherway.

**Comment by yezz123** [comment]
> > hey @Youssefares correct me if i'm wrong but i guess its similar to this one we have in Pydantic-extra [pydantic/pydantic-extra-types#193](https://github.com/pydantic/pydantic-extra-types/pull/193)
> 
> This does seem very similar, zoneinfo.ZoneInfo is part of the python std library though. I maybe wrong but my understanding of pydantic-extra-types is that it defines additional convenience types, while pydantic itself is where we offer support for standard types. I am not 100% sure about the need for an extra-type in this scenario given zoneinfo now exists since 3.9. I also think (in my experience at least), most people try to drop the pytz dependency from their project nowadays if zoneinfo does the job.
> 
> I'll leave to maintainers to decide whether the extra type using pytz is still required, but I believe supporting zoneinfo is very much worthwhile eitherway.

Yes as you said the best answer we can have is from maintainers themselves.

cc @sydney-runkle @samuelcolvin

**Comment by Youssefares** [comment]
please review

**Comment by sydney-runkle** [comment]
Exciting stuff here - this is next in my queue to review :). Should be able to get to it this evening or first thing tomorrow :).

File: pydantic/_internal/_generate_schema.py [hunk 1]
```diff
@@ -1473,6 +1475,31 @@ def _type_schema(self) -> core_schema.CoreSchema:
...
+            try:
+                return ZoneInfo(value)
+            except (ZoneInfoNotFoundError, ValueError):
+                raise PydanticCustomError('zoneinfo_str', 'invalid timezone: {value}', {'value': value})
+
+        ser = core_schema.plain_serializer_function_ser_schema(lambda instance: str(instance), when_used='json')
+        metadata = build_metadata_dict(js_functions=[lambda _1, _2: {'type': 'string', 'format': 'zoneinfo'}])
+        return core_schema.no_info_plain_validator_function(
+            validate_str_is_valid_iana_tz,
+            serialization=ser,
```
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply pydantic/_internal/_generate_schema.py:1499]: `core_schema.to_string_ser_schema()`

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Very excited to include this. Added one minor simplification request!

**Comment by sydney-runkle** [comment]
> I'll leave to maintainers to decide whether the extra type using pytz is still required, but I believe supporting zoneinfo is very much worthwhile eitherway.

I think we should definitely just stick to std lib support, so the zoneinfo addition here is great!

**Comment by Youssefares** [comment]
CI's failing because this issue #9761 been resurrected somehow it seems? Looking into it..

**Comment by Youssefares** [comment]
That look better.. 🟢 ✅

File: pydantic/_internal/_generate_schema.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Youssefares) [reply pydantic/_internal/_generate_schema.py]: Thanks! I've addressed this.

**Review by sydney-runkle — APPROVED** [review]
Looks great!

Thanks a bunch!

**Pull request merged**